serialize/src/impls.rs

```diff
@@ -458,13 +459,21 @@ impl<T: CanonicalDeserialize, const N: usize> CanonicalDeserialize for [T; N] {
         compress: Compress,
         validate: Validate,
     ) -> Result<Self, SerializationError> {
-        let result = core::array::from_fn(|_| {
-            T::deserialize_with_mode(&mut reader, compress, Validate::No).unwrap()
-        });
+        let mut array = ArrayVec::<T, N>::new();
+        for _ in 0..N {
+            let elem = T::deserialize_with_mode(&mut reader, compress, Validate::No)?;
+            // Defensively use `try_push` error to ensure that we *never* panic:
+            array
+                .try_push(elem)
+                .map_err(|_| SerializationError::NotEnoughSpace)?;
+        }
         if let Validate::Yes = validate {
-            T::batch_check(result.iter())?
+            T::batch_check(array.iter())?
         }
-        Ok(result)
+        Ok(array
+            .into_inner()
+            // Defensive error return:
+            .map_err(|_| SerializationError::NotEnoughSpace)?)
```

Comment:
can this line ever throw an error? It looks like `NotEnoughSpace` should have been caught in the loop above

Reply:
It should not, which is why i commented `Defensive`.  
I can also unwrap it if you prefer, but then `T: Debug` is required.

Reply:
It [does](https://doc.rust-lang.org/std/result/enum.Result.html#method.expect)... but i can use  an `.unwrap_or_else(|_| unreachable!())`.

Reply:
Oh I was looking at `Option::unwrap()` lol

Reply:
`.map_err(|_| ()).unwrap()` works too.